### ISSUE TITLE
Suggestion: collapse unchanged regions in diff

### ISSUE THREAD

**Issue by factormystic** [Microsoft/vscode#3562]
When viewing a diff (2 pane or inline) it would be useful to collapse the unchanged lines. This helps focus in on just the changes, especially when the changes are few an spread out.

**Comment by bkarlson** [comment]
any update on this one? Very hard transition from text-based git to VSCode because of this feature lacking.

**Comment by bkarlson** [comment]
bump...

**Comment by benlindsay** [comment]
This would be very nice to see. Right now, with big files I'm opening Vim (with fugitive plugin) in the terminal to clean up the diffs

**Comment by Rijk** [comment]
This is something I can't live without, the collapsing along with the inline Stage/Discard buttons makes it so much easier to keep commits tidy:

![image](https://user-images.githubusercontent.com/159500/79044299-9c2b0580-7c04-11ea-996b-ffdcfac1c58c.png)

This screenshot is from [GitX](https://rowanj.github.io/gitx/) — still by far my favorite Git UI. I have tried literally all the extensions in VS Code, still none comes close.

**Comment by Ardeshir81** [comment]
no update on this issue?
any one knows any extension?

this issue is open since 2016

**Comment by ryran** [comment]
VS Code could really stand to make some huge improvements by looking at Atom's git support -- viewing diffs, walking through changes & staging them, as well as undo/reset ... it really is quite a bummer that these features are lacking.

**Comment by KayLeung** [comment]
Yeah. To hide unchanged context is a very common feature for all diff editor and I think the MS developers also need this feature for each commit. I wonder why they abandoned the implementation for such a long time.

VSCode has built-in collapse/expand feature for code editor. A different algorithm would make the deal it seems.

**Comment by jianboy** [comment]
four year..... the function is still missing.

**Comment by cristian3d** [comment]
This is, by far, my most wanted feature. It is standard in many other tools, specially when doing source control diffs. 

I'd appreciate some comment from the team as to why is not being considered. Is it too hard to implement?

**Comment by andronat** [comment]
The only reason I'm still using GitHub Desktop is because of this missing functionality. I'm often working on large files of C/C++ with 1000s of lines of code and long scrolls are just killing me.

Please make sure that the "selecting ranges to Stage" feature won't break!

**Comment by ToMakeSense** [comment]
related to #87944

**Comment by sdegutis** [comment]
This is one of the main things Magit gives me and Magit is the only reason I use Emacs anymore, please add this feature.

**Comment by jakedowns** [comment]
i know i'm not supposed to comment just to 👍 but here i am commenting simply to 👍 
anyone know of a plugin to do this instead?

**Comment by ADTC** [comment]
@jakedowns Try using the "vscode-diff" extension. It doesn't solve the problem. However it generates and open the output of `git diff` in VS Code.

Do note that if you're on Windows, you need to modify the TS file of the extension to fix the problem described in huytd/vscode-diff#1

**Comment by saxicek** [comment]
It is a shame that `git diff` provides better overview of changes than VS Code. 👎🏻

**Comment by HebaruSan** [comment]
Maybe this is waiting for community contributions. Is this the right file?

https://github.com/microsoft/vscode/blob/main/src/vs/editor/browser/widget/diffEditorWidget.ts

**Comment by archon810** [comment]
> It is a shame that `git diff` provides better overview of changes than VS Code. 👎🏻 

Not just git diff. Github itself, Bitbucket, Beyond Compare, pretty much all code diff tools do this.

**Comment by ThomasGHenry** [comment]
I left VS Code for WebStorm a few years ago. I thought I might come back, but I keep bumping into things like this.

**Comment by jordyvandomselaar** [comment]
Hi @archon810 nice bumping in to you 😁 

Agree, for now I just use the "jump to next change" button to work around this, but the current diff viewer makes for a cluttered experience.

**Comment by rijk** [comment]
Cluttered and a little stressful, because I never feel 100% sure I've seen everything.

**Comment by FarisHijazi** [comment]
this would definitely be a much needed feature

Something similar to Intellij diff as shown bellow:

![](https://i.imgur.com/lhBrIDp.png)

**Comment by ivanschuetz** [comment]
Please..? I've tried to get used to this, but with collapsed regions the UX is just much better. I keep going back to Sourcetree just because of this.

**Comment by vsambor** [comment]
The problem is that the folding does not even work inside a diff file. 
In my case I would like to get a screenshot for a small diff, the diffs are in the same file, but in different far locations, so it's not possible to fold irrelevant code, or to have some option to show only the diffs as in other tools.

**Comment by thomashbrnrd** [comment]
This is currently the only reason why I got some trouble to move from Atom to VS Code :(

**Comment by ADTC** [comment]
@thomashbrnrd I just use another tool for diff. VSCode is better than Atom so this one thing shouldn't stop you.

(Also I believe Atom is retired.)

**Comment by thomashbrnrd** [comment]
Yes Atom is retired, but Atom was really better than VS Code on the versioning / git side. 
This feature that we request here was implemented long time ago in Atom ...

**Comment by ADTC** [comment]
@thomashbrnrd I prefer to use Git in terminal command line (external) and visually in another tool. In VS Code I use some Git related extensions but only to get information about the current history.

If you're on Windows, you can try Git Extensions. It's very good and can do line by line staging. It's also possible to use it on Mac with some strange emulation but I didn't want that so I switched to Fork on my new Mac.

**Comment by antichris** [comment]
@thomashbrnrd If you're comfortable with text mode tools, you could use interactive add (`git add -i`).

If you prefer GUI and you're fine with the tool not being exceptionally pretty, I'd recommend the official [`git-gui`], which comes out of the box with Git on some platforms (on many Linux distros you'd have to manually install the corresponding package, typically, `git-gui`). It's what I use and it works reasonably well.

It really does suck that it has taken so long for VSCode to still have no support for hiding/collapsing unchanged diff regions. Evidently, garbage like "[custom Explorer autoreveal logic](https://code.visualstudio.com/updates/v1_74#_custom-explorer-autoreveal-logic)", "[hide Activity bar and Panel badges](https://code.visualstudio.com/updates/v1_74#_hide-badges-per-view-container)" or "[new font setting for OpenType fonts](https://code.visualstudio.com/updates/v1_74#_new-font-setting-for-opentype-fonts)" is more essential to the development workflow for some; since this issue was opened almost 7 years ago there have been tens of releases with hundreds of superfluous insignificant features like those. :disappointed: 

[`git-gui`]: https://git-scm.com/docs/git-gui

**Comment by mtazzari** [comment]
How is it possible that this hasn't been done in 6 years! Much needed feature to be competitive with, e.g., PyCharm.

**Comment by aviallon** [comment]
Why is this still a missing feature?

**Comment by ADTC** [comment]
> Why is this still a missing feature?

To encourage us to use `git diff` or other tools so we don't miss this feature as much.

Personally I use Fork now. I used to use Git Extensions. Sometimes I use Beyond Compare for more complex diffs that Git can't properly show changes on.

**Comment by fr-an-k** [comment]
This is essential; what is going on?

**Comment by situ-s** [comment]
This is very much needed when there are lots of untracked files generated via scripts. Any update?

**Comment by Rocamonde** [comment]
Lol how is this not implemented?

**Comment by casnix** [comment]
This would be super useful, even outside of git tracked repositories (for e.g. folders).  Maybe we have not given enough thumbs up to have this added?

**Comment by llu025** [comment]
I'll be posting a comment every day until this is solved, this is absurd

**Comment by antichris** [comment]
@llu025 

> I'll be posting a comment every day until this is solved, this is absurd

Don't. You're not five, you understand that all that's going to achieve is annoy the hell out of the people who have subscribed to this issue until getting yourself blocked.

**Comment by FarisHijazi** [comment]
How is this not solved yet? Do we need to get more thumbs ups? Do we need to open new issues eith the same request? What do we need to do to get done?

Share this with friends so they can thumbs up

**Comment by rijk** [comment]
While I also **really** want this feature (it's the only part of my coding flow I still use another app for), I feel adding "why is this not implemented yet" comments is not constructive. I also frankly find it a bit strange to act so entitled about a piece of free software that no one forces you to use, and make demands of the maintainers. 

By the way, I believe that VS Code is open source for you to contribute to if you wish to spend the time on this feature yourself.

**Comment by Rocamonde** [comment]
@rijk I disagree. Adding "why is this not implemented" comments reflect the actual demand for the feature, and is extremely useful information for maintainers, whose goal is to prioritize what features should be worked on. This should not be conflated with being rude or unfriendly, which should be discouraged.

Expecting everyone to either contribute through a PR or shut up is naive and toxic behavior. Demands from the community, feedback forums, etc. are the signals that OSS has to decide on the allocation of resources and human capital, since there's no price signal unlike commercial software.

**Comment by Phoenix1747** [comment]
> While I also **really** want this feature (it's the only part of my coding flow I still use another app for), I feel adding "why is this not implemented yet" comments is not really helping. Maybe we should relax this discussion a bit, and realize we are not entitled to making demands of the maintainers. VS Code is a piece of free software that no one forces you to use, and you're free to contribute if you wish to spend the time on this feature yourself.

Asking "what do we need to get this done" has nothing to do with being entitled. If feature requests and the respective community demand were unwanted then this whole thread wouldn't exist. Yet here we are seeing that many people would love this feature. No more, no less.

**Comment by antichris** [comment]
@Rocamonde

> toxic behavior

That choice of words is... _interesting_.

No, friend. It is a well-know and very common convention in many virtual social spaces — that one can like/thumbs-up/"heart"/upvote/etc. to signal their reaction to a post/thread/comment/issue without having to squirt a worthless "+1"/"same here"/"me too!" in the discussion. In fact, many community guidelines _explicitly_ advise against, if not outright prohibit the latter.

One's "me too!" comment adds exactly zero _new_ feedback to the discussion that couldn't have been added by a simple and polite upvote (or downvote, if that's the sentiment). It just rudely wastes everyone else's digital and mental bandwidth. Including the developers, who might have to scroll through screenfuls of that sort of crap in a vain effort to find any actually useful feedback.

Please, don't post if you don't have anything to say that hasn't been said already. Upvote :+1: the issue instead!

**Comment by ivanschuetz** [comment]
Just noting - I just learned that this exists, that you can sort the issues by upvotes: https://github.com/microsoft/vscode/issues?q=is%3Aissue+is%3Aopen+is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc+

We're somewhere in the second page now.

**Comment by ivanschuetz** [comment]
> Just noting - I just learned that this exists, that you can sort the issues by upvotes: https://github.com/microsoft/vscode/issues?q=is%3Aissue+is%3Aopen+is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc+
> 
> We're somewhere in the second page now.

It's probably worth noting too that the most upvoted issue is also from 2016, and was locked in 2018, mentioning that it's "on the team's radar" and to just upvote it, so questions about how the team is prioritizing things are not entirely inappropriate..

**Comment by rijk** [comment]
No, I agree. As @antichris [mentioned](https://github.com/microsoft/vscode/issues/3562#issuecomment-1365169416) as well, it's somewhat puzzling to see which features do appear on the changelog each month and not getting even a response by the team on this one. Are they even looking at Github issues for determining the feature roadmap? 

I found [this document](https://github.com/microsoft/vscode/wiki/Roadmap) outlining their overall roadmap, and they work off of monthly "iteration plans" [like this one](https://github.com/microsoft/vscode/issues/176147).

So the question is: how do we get on the next one? 🤠 It seems @lszomoru is working on the source control related issues, maybe we can get his take on this suggestion?

**Comment by rasborgdk** [comment]
Trying to make sure that the EDI is getting setup - done correctly

**Comment by adam-rocska** [comment]
pls pls pls

**Comment by rijk** [comment]
They're working on it! (it's on the April iteration plan) 🙌

We made it peeps

**Comment by sashahilton00** [comment]
> They're working on it! (it's on the April iteration plan) 🙌
> 
> We made it peeps

Had to check this wasn't an April fools...

**Comment by orthodoX** [comment]
Well good luck in the next release everyone, cause it's not in this one 😅

**Comment by hediet** [comment]
Very early prototype:

![Code_-_OSS_R7MCPBUAob](https://github.com/microsoft/vscode/assets/2931520/a3e45f9d-e725-4baf-a157-53dadebdf816)

Because this feature is part of a bigger diff editor refactoring, we cannot give any timeline on when this feature ships.

**Comment by ADTC** [comment]
@hediet That looks very cool. Can we see a version of it in inline diff mode? I don't use side by side mode.

If I may suggest something, I think the wavy border is too loud. A plain straight border should be enough, as the color alone is good enough to distinguish the bar from the code.

**Comment by cben** [comment]
Also, with [sticky scroll](https://code.visualstudio.com/updates/v1_70#_editor-sticky-scroll), vscode already has a very subtle way to show non-contiguous lines, taking up 0 vertical space.  
And block folding >v hides lines with an ellipsis + background color for the line above.  
![bildo](https://github.com/microsoft/vscode/assets/273688/acde4c17-6dab-4261-a4a1-381c96852dac)
You might want to consider some visual unity between those 3 scenarios...

However, we'll take this feature gladly with any styling whatsoever :heart_eyes:

**Comment by lobsterkatie** [comment]
> If I may suggest something, I think the wavy border is too loud. 

Agreed. It looks as if someone has taken pinking shears to the code, LOL.

But as others have said, thanks for your work on this, and looking forward to it shipping!

**Comment by hansu** [comment]
The older IntelliJ version could be inspired from the "broken view" representation of mechanical drawings like this:
https://help.solidworks.com/2016/english/solidworks/sldworks/c_broken_view.htm

But there are also other variations possible which might give further inspiration:
https://help.solidworks.com/2021/english/SolidWorks/sldworks/c_broken_view_break_line_styles.htm

**Comment by ADTC** [comment]
Just a note that I prefer inline diff (not side-by-side), so it should also work well on that. It seems the discussion is biased to side-by-side views 🙂

**Comment by hediet** [comment]
Here is a negative example of SmartGit, where I initially thought that `setBoundarySashes` was added to `LineRangeMapping` (which is not the case):

![image](https://github.com/microsoft/vscode/assets/2931520/be67e67d-b606-4cfa-9646-81086e619472)

**Comment by ADTC** [comment]
@hediet since we already have sticky method names (etc), I think the same can be applied to the diff bar, which will help mitigate the negative example above. I think `git diff` and GitHub do this.

**Comment by hediet** [comment]
See [here](https://github.com/microsoft/vscode/pull/183535) for how you can try out this feature in the new very experimental rewrite of the diff editor (available in next weeks insider). Please note the new diff editor is not production ready yet.

More polishing will happen in June.

**Comment by Rocamonde** [comment]
Great work guys, thank you for this!

**Comment by EmrecanKaracayir** [comment]
There is a theme attribute called `diffUnchangedRegionBackground`. How can we see this unchanged region attribute? I activated diff editor v2

**Comment by hansu** [comment]
Great work! 
But the directions of the arrows are a bit irritating.
When clicking on `^ Show More`, I would expect to get more lines of the code part that is below that button like it is implemented in GitHub, not the lines above.

![bild](https://github.com/microsoft/vscode/assets/67957916/473a9adf-b266-4872-a12a-9e4be582673b)

![bild](https://github.com/microsoft/vscode/assets/67957916/e9409df3-e127-4f39-88f7-c21907108a8d)

**Comment by hansu** [comment]
>How can we see this unchanged region attribute in action? I activated diff editor v2.

@EmrecanKaracayir  I agree, the collapsed region could be highlighted a bit more - with the current coloring it will be overlooked most times

**Comment by ADTC** [comment]
> But the directions of the arrows are a bit irritating.
> When clicking on `^ Show More`, I would expect to get more lines of the code part that is below that button like it is implemented in GitHub, not the lines above.

Please reverse the arrows. In GitHub the up arrow ⬆️ means move this collapsed code bar UPwards and reveal more code below it. That makes sense and it's what we're used to.

**Consider using "⬆️ Expand Up", "⬇️ Expand Down" and "↕️ Expand All" which is the same as the tooltips shown in GitHub!**

**Comment by hediet** [comment]
Thanks for the valuable feedback! I agree we should try to align with GitHub.

**Comment by cben** [comment]
I'd say a mere arrow/triangle, at middle height is inherently ambiguous.  (e.g. say it points up ⮝)
1. it could mean "show more code _on that side_ of the break (above) 
2. it could mean "expand the visible code in this direction (_intruding upwards_ into the break from below).

Note how SmartX positioned the triangles 🢑low / high🢓 to reinforce the second interpretation (e.g. 🢑 is pointing up, but it will move the bottom boundary of the break).
And how GitHub's icons reinforce the same by literally showing a dashed boundary that would be "dragged":
![github screenshot](https://github.com/microsoft/vscode/assets/273688/ca1c72eb-38bd-43b8-b9b0-d8598cba365e)

Well the Show all icon in above VSCode screenshots did disambiguate them :+1: — it'd be something like ↕ for 1st interpretation "show more on both sides" vs. the X suggesting the 2nd "drag both boundaries inwards until they meet".  But it takes some thinking, and would work better if the one-sided icons were aligned at similar heights.  
GitHub's "expand all" icon actually mixes the metaphor :-1:, by above logic it should mean "drag both boundaries outward, showing _less_ code"?!  Maybe don't copy that ;-)

GitLab also has the same mixed metaphor for diff breaks, with "outward arrows for "Expand all lines" (but using inward arrows for collapsing a comment thread — not exactly same thing but at least makes sense to me)
![gitlab screenshot](https://github.com/microsoft/vscode/assets/273688/dc9d4397-d196-44b6-bd24-e74249660b2a)
(Am I missing some third interpretation by which it's all consistent?)

As to the tooltips, I'd say by themselves "Expand up"/"Expand down" are still ambiguous; they only work in GitHub because the icons are stacked vertically, leaving no doubt which boundary would move.  "Show more below" or "Expand upward" might be clearer?

**Comment by hediet** [comment]
> by literally showing a dashed boundary that would be "dragged":

This interpretation is a bit confusing, as the height of the diff break stays constant.

Thinking about it, I actually prefer the terminology "Show more below/above", as this cannot be misinterpreted. With this terminology the arrow directions are correct.
Personally, I don't find "Expand upwards/downwards" as clear - it could also mean that the diff break is expanded.

I agree that other tools solve this by stacking the buttons vertically. Maybe we should also do this, even though it wastes some space.
I also don't want to include the method name in the diff break instead, but rather break up the diff breaks so they reveal the method/class headers - for this, a smaller diff break would be beneficial.
Additionally, I think the show above/below buttons should only show on hover.

**Comment by cristian3d** [comment]
I know this is a work in progress, but "Stage Selected Ranges" doesn't work for me when using the new diff. It says "The selection range does not contain any changes.". Other than that, I'm excited to finally have the collapsed diffs! Cheers!

**Comment by hansu** [comment]
>Thinking about it, I actually prefer the terminology "Show more below/above", as this cannot be misinterpreted.

I also prefer that. But the arrow directions...

>Also, I think the show above/below buttons should only show on hover to reduce the visual noise.

If showing them only on hover, they could the displayed over the squiggly border to save space:
![grafik](https://github.com/microsoft/vscode/assets/67957916/746140c0-0c46-4113-9f6b-d90f3fe96d78)

**Comment by hediet** [comment]
> I know this is a work in progress, but "Stage Selected Ranges" doesn't work for me when using the new diff. It says "The selection range does not contain any changes.".

Yes, this is not implemented yet. Thanks for trying out the new diff editor already though! With the next insider there is also the inline mode and moved code detection.
See [here](https://github.com/microsoft/vscode/issues/184164) for what already works in the new diff editor.

**Comment by hediet** [comment]
Some UI update (expect more changes):

![Code_-_OSS_y0vYGOMXFR](https://github.com/microsoft/vscode/assets/2931520/535eba26-b138-4162-99f7-8c1f9a438edf)

Also, I think some drag&drop of the borders to "pull out" hidden code would be an intuitive gesture to show more.

**Comment by chazmo03** [comment]
The "Show All" functionality is not "remembered" when navigating back and forth between tabs, which is confusing. It would also be nice to have a "Show full file" option somewhere to just expand everything.

**Comment by ADTC** [comment]
> I think some drag&drop of the borders to "pull out" hidden code would be an intuitive gesture to show more.

That will be nice. Dragging the bar up or down could show more code smoothly line by line as much as needed without having to click show more multiple times. A drag handle could be added at the left edge. Dragging up to show more below, then immediately down (without releasing) could hide previously shown lines. (And vice versa.)

 For consistency the arrows should match the direction of dragging: "⬇️ (drag down to) Show more above, ⬆️ (drag up to) Show more below". Right now, the direction doesn't match the actual direction towards which the bar would move.

**Comment by hediet** [comment]
Another idea (ignore the lagging and flickering, it's a prototype to evaluate the idea):

![lBoGyqU0TZ](https://github.com/microsoft/vscode/assets/2931520/adec689d-4d78-465f-972a-8786a85e630a)

**Comment by rijk** [comment]
Oo that dragging looks really nice!

**Comment by dimateos** [comment]
> Another idea (ignore the lagging and flickering, it's a prototype to evaluate the idea):

Looks excellent! I would prefer that along with the mentioned ↕️ symbol for expanding (the x seems weird).

**Comment by hansu** [comment]
The dragging feature looks nice, but I am not sure if it will prove itself in practice. Clicking "Show More" is quicker and doesn't need that exact positioning of the mouse pointer.

**Comment by LPTK** [comment]
@hediet 

> Another idea

Looks cool!

But the fact that the dragging gesture can hide what was revealed earlier doesn't seem useful.

How often does one need to hide something they've just revealed when looking at a diff? Given the fact that UIs like GitHub don't have a button for that, I guess the answer is basically never.

So the dragging could be performed by clicking _anywhere_ on the black separation space, and while dragging, moving the mouse up should reveal more of the code below, and moving down should reveal more of the code above.

This way, no need to precisely click on one of the two upper or lower edges of the bar.

**Comment by dimateos** [comment]
I wonder if there is going to be a setting for the default amount of lines displayed around unchanged regions and the amount revealed by "show more".

I feel like the current `3` (before/after) may not provide enough context for most cases, i.e. function body, loop, etc.

This feature is great for large files where you can hide sizeable chunks of lines. E.g. I want to collapse the first 300 untouched lines, but then I prefer like 15 before/after for context.

Further more, IMO this setting would also leave a bit more breathing room between these new black stripes. They can be a bit distracting. I hope there is a `"workbench.colorCustomizations"` to reduce the contrast.

**Comment by antichris** [comment]
> ![lBoGyqU0TZ](https://user-images.githubusercontent.com/2931520/246198866-adec689d-4d78-465f-972a-8786a85e630a.gif)

I kind of like this UI. I think the active zones of the divider should be not just the very edges, but rather the top and bottom halves. Dragging could gradually reveal lines like in the prototype, clicking should reveal a fixed number of lines (5, 10, 15, or, optimally, configurable). A tooltip could help with discoverability. The exact appearance of the divider is debatable though, I would prefer a straight edge with a hint of shadow, similar to how the sticky scroll feature is done.

I think the number of context lines shown initially should have a global setting (user, workspace), that can be adjusted per file (ideally, remembered, but can be reset back to the configured value).

And there should be an action (or two) to collapse/expand all unchanged regions in the current diff view.

**Comment by ADTC** [comment]
> The exact appearance of the divider is debatable though, I would prefer a straight edge with a hint of shadow, similar to how the sticky scroll feature is done.

To me there's no debate. The wavy line looks cartoonish in an otherwise professional user interface.

Is no one else against it? 😅

**Comment by antichris** [comment]
> Is no one else against it?

I wouldn't say I'm _terribly_ happy with the way it looks now. But I'm also one of those who still use the official Tcl/Tk `git gui` in 2023, so it may be a very reasonable guess to say that I don't really give two shits about pretty graphics when it comes down to tools. As long as it's not so bad that it gets in the way of what one is trying to do.

I just noticed that the draggable prototype seems to snap to whole lines. I may be in minority, but I think it should only snap when the exact number of context lines is changed (e.g., on a click to show more lines). Dragging should be smooth, even if it could end up cutting off a part of the line when let go. UI components snapping out and away from under the cursor, on the other hand, _do_ make for a very lousy UX — one that can get in the way of what someone is trying to do.

**Comment by hediet** [comment]
Here another idea:

![Voe6UTbQKf](https://github.com/microsoft/vscode/assets/2931520/fdc6b622-7037-492c-b0e3-dfd368aa6f0c)

The drag&drop idea can be implemented in the future on top of this design (which is technically quite involved to do it properly, so it probably won't be included in the initial release of the new diff editor)

This prototype also shows how context could be included, similar to what GitHub does (current method/class).
Initially I thought the header lines could just stay visible (surrounded by the break-spacer), but that adds too much noise.

**Comment by ADTC** [comment]
Loving the new clean straight-edged bar. Perfectly blends in with the UI while standing out just enough to notice without distracting. Thank you so much!

I think I'm also in agreement with the new version, to click the border. Could you possibly make the click target larger though, at least 5 pixels (inside the bar at the border), even if the border itself is only 1-2 px?

Then I have one more suggestion, optional but for consideration: instead of the standard tooltip that opens after a delay, consider replacing the text "32 hidden lines | 🎏 DiffEditorWidget2" with "Show more lines above/below" and "Show all lines" when hovering the respective click target, then return it back to the original text when mouse is moved away or when "Show more lines above/below" is clicked (updating the number of lines).

**Comment by antichris** [comment]
> Could you possibly make the click target larger though, at least 5 pixels (inside the bar at the border), even if the border itself is only 1-2 px?

I still stand behind the idea that the entire half of the bar should be the clickable active zone. Here's a quick mockup:
![image](https://github.com/microsoft/vscode/assets/881420/69c0f92a-5dd5-4661-92be-f2e920f193ea)


<details><summary>In this case it may also be worth considering making the gutter with the "Show all" button not covered by the active zones. (Expand for a picture.)</summary><p>

![image](https://github.com/microsoft/vscode/assets/881420/f6df0abc-a68f-47a3-8ade-97a09d213d14)

</details>

**Comment by hansu** [comment]
> Here another idea:

No, please don't do that. It's very un-intuitive that there are borders which are suddenly clickable. Further it needs precise pointing of the mouse to hit that thin line.

I think reserving the space of two lines with one button on top of the other is totally fine. 
Keep in mind that you should compare the space actually needed by the bar to the space of the lines that are collapsed and not thinking that it is an additional bar that just wastes you space.

**Comment by ADTC** [comment]
@hediet I'd say that hansu has a good point. Using a border to do something other than changing the width/height of a section is _not_ normally expected. It will be a brand new behavior to learn from a UI/UX perspective - something we should avoid adding unless we're introducing a groundbreaking new UI behavior with many potential use cases. 

It's OK for a collapsed code bar to take up about `1.75` to `2.0` lines of height, (especially now that you've gotten rid of the wavy edge which took up too much space). I think the GitHub style is good enough, which shows expand up/expand down buttons stacked. Maybe holding Control or Option(Alt) could change the split buttons to a single Show All button (which is also normally shown when there's not enough lines to warrant the split buttons).

We should also have a separate option somewhere to expand and collapse all sections of unchanged regions in the file. It's something GitHub has in the file header (because the diff page has many files in it). [Note: Although GitHub remembers the individually expanded sections and keeps them expanded when we collapse all sections across the whole file, this isn't strictly necessary.]

Please give it a try, and show us a demo when you can. Thank you.

**Comment by ADTC** [comment]
GitHub screenshot for reference:
<img width="557" alt="image" src="https://github.com/microsoft/vscode/assets/6047296/8161dec9-5164-48a4-800d-46f786dd7784">

**Comment by ADTC** [comment]
There's also the use of a different background color and a subtle border in GitHub for the expanded parts of the code, which is also nice to have, as it distinguishes the unchanged code (outside context limit) from the changed code and the limited context around it.

<img width="397" alt="image" src="https://github.com/microsoft/vscode/assets/6047296/c6cb517f-c34b-47e5-9482-5d385ea194a9">

**Comment by rijk** [comment]
Awesome!! 🕺

**Comment by hansu** [comment]
Great!
But feels still weird with the arrow directions (contrary to GitHub) especially in this case:
![Peek 2023-06-27 22-23](https://github.com/microsoft/vscode/assets/67957916/eb5ccb33-50ac-4fb8-a396-16e05d2327b4)

**Comment by hansu** [comment]
I further wonder if it makes sense to collapse unchanged regions when having only one change. 
The diff editor V1 scrolls automatically to the change so there is no real need for collapsing and it might be more convenient to show the whole file in this case.

**Comment by rijk** [comment]
I don't agree with that. Collapsing unchanged code is always a good thing to reduce mental tax and ensure you've seen all the changes. Regardless if 1 or multiple changes.

**Comment by hediet** [comment]
@hansu this is an old version. Please try the latest insiders!

**Comment by hansu** [comment]
>I don't agree with that. Collapsing unchanged code is always a good thing to reduce mental tax and ensure you've seen all the changes. Regardless if 1 or multiple changes.

Maybe this comparison makes it a bit more clear:

Diff editor v1:
![bild](https://github.com/microsoft/vscode/assets/67957916/c3bbf2f8-bcef-4b02-b4a7-528b96365908)

Diff editor v2:
![bild](https://github.com/microsoft/vscode/assets/67957916/f24f1f55-8dd5-4692-a9e7-d7b9fe8cd63b)

**Comment by ADTC** [comment]
Why is almost everyone sticking with side-by-side view? 😂😂😂 I find inline view to be much better. Full 100% width, and no redundancy of seeing the same unchanged line twice. I hope this feature works well with inline view as I haven't seen any screenshot of that yet.

**Comment by hansu** [comment]
>@hansu this is an old version. Please try the latest insiders!

Oh sorry, I didn't check that.
But the current version is even more unconvenient in this case:
![diffeditorv2-1](https://github.com/microsoft/vscode/assets/67957916/4210b165-1514-4264-b1ad-1a74ff71024c)

Also a bit weird that the text goes first down and then up. I hope you get back to the buttons you had before :)
![diffeditorv2-2](https://github.com/microsoft/vscode/assets/67957916/c030d2cc-7732-4cad-be4a-3c49496e9142)

**Comment by hediet** [comment]
> Also a bit weird that the text goes first down and then up. I hope you get back to the buttons you had before :)

That is a bug. These sashes are actually clickable and do the same as the button!

**Comment by hansu** [comment]
>That is a bug. These sashes are actually clickable and do the same as the button!

Oh, I didn't expect that.

**Comment by rijk** [comment]
<img width="1525" alt="image" src="https://github.com/microsoft/vscode/assets/159500/e0c8c097-9b76-4aed-8233-bdad057727df">

To me, it's perfect! 😍 Great work @hediet

**Comment by KayLeung** [comment]
Finally!!! This landed on July 2023 (version 1.81) (experimental)!!!!!

Only one problem, the changed view has inline hints. It increases the noise to compare 2 sides.
![Screenshot 2023-08-04 030646](https://github.com/microsoft/vscode/assets/32264/7431d274-9fb0-42aa-87ec-1c7332a123d5)

**Comment by ADTC** [comment]
Could anyone please share screenshots / videos of how the inline (unified) diff looks like? Thank you.

I'm afraid to switch because I feel like I'll be disappointed. All the talk so far have been about perfecting the side-by-side, and I only use inline diff. :)

**Comment by yume-chan** [comment]
There are some issues in the inline view:

With `"editor.lineNumbers": "off"`, there are two collapse buttons after expanding collapsed area:

![image](https://github.com/microsoft/vscode/assets/1330321/20406453-e378-463e-8b19-6f927539f1ac)

(I think it doesn't need two even when `"editor.lineNumbers": "on"`, as it only have one "Show All" button)

Switching `editor.lineNumbers` on/off, the layout is broken, need to resize the editor to let it re-render:

Switch from "off" to "on":

![image](https://github.com/microsoft/vscode/assets/1330321/45f2123c-d46a-473c-856d-b6c63bbfc1b7)

(Only shows one column of line numbers)

From "on" to "off":

![image](https://github.com/microsoft/vscode/assets/1330321/9025ebde-4032-4d5d-b69a-5a4c8b3ab1dd)
